Task 2

Pl-Alex opened the pull request.

Pl-Alex committed:
Level 2, 1 task

Pl-Alex committed:
tast-2 deleting a contact

ADDED:
- deletion of contacts by pressing a button and swipe with the ability to cancel the action

Pl-Alex committed:
task-2 Added add contact dialog

CHANGED:
- DataStore transferred to a separate file with data access functions
- Added appbar in activity_contacts.xml
- Some code refactoring
ADDED:
- Adding a contact using the dialog
- DiffUtils implementation

Pl-Alex committed:
CHANGED:
- Migration from RecyclerView adapter to ListAdapter

Pl-Alex committed:
task-2 recyclerView bug fixing

fixed:
- Scroll to the inserted item
- Spacing of the first item in the list

Pl-Alex committed:
task-2 Receiving contacts from a device

ADDED:
- Receiving contacts from a device

Pl-Alex committed:
task-2 some refactoring

FIXED:
- animation bug when adding element to RecyclerView
CHANGED:
- Refactoring of validation

Pl-Alex committed:
task-2 Fixed app theme

FIXED:
- Light and dark theme
- default_contact_image icon

Pl-Alex committed:
task-2 Added GoogleButton CustomView

ADDED:
- GoogleButton CustomView
- ripple effect to GoogleButton CustomView
- Additional attributes to GoogleButton CustomView

Pl-Alex committed:
Merge remote-tracking branch 'origin/master' into dev

# Conflicts:
#	app/src/main/java/com/alexP/assignment1/ui/AuthActivity.kt
#	app/src/main/java/com/alexP/assignment1/viewModels/AuthViewModel.kt
#	app/src/main/res/layout/activity_auth.xml
#	app/src/main/res/values/dimens.xml

Pl-Alex committed:
Merge remote-tracking branch 'origin/master' into dev

# Conflicts:
#	app/src/main/java/com/alexP/assignment1/ui/AuthActivity.kt
#	app/src/main/java/com/alexP/assignment1/viewModels/AuthViewModel.kt
#	app/src/main/res/layout/activity_auth.xml
#	app/src/main/res/values/dimens.xml

Pl-Alex committed:
Merge remote-tracking branch 'origin/task-2' into task-2

# Conflicts:
#	app/src/main/res/values/dimens.xml

1937Carrie commented on gradle/libs.versions.toml:
```diff
@@ -24,4 +26,6 @@ junit = { group = "junit", name = "junit", version.ref = "junit" }
 datastorePreferences = { group = "androidx.datastore", name = "datastore-preferences", version.ref = "datastore" }
+cardView = { group = "androidx.cardview", name = "cardview", version.ref = "cardview" }
+javaFaker = { group = "com.github.javafaker", name = "javafaker", version.ref = "javafaker"}
 [plugins]
```
There is empty the plugins part. Look [this](https://developer.android.com/build/migrate-to-catalogs#migrate-plugins) or [this](https://youtu.be/MWw1jcwPK3Q?feature=shared).

1937Carrie commented on app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    fun loadContacts() {
+        if (ContextCompat.checkSelfPermission(
```
Using onRequestPermissionsResult() in AppCompatActivity is indeed possible ([proof](https://developer.android.com/reference/androidx/core/app/ActivityCompat.OnRequestPermissionsResultCallback#onRequestPermissionsResult(int,java.lang.String[],int[]))). But prepare to level 3 and read [this](https://developer.android.com/reference/androidx/fragment/app/Fragment#onRequestPermissionsResult(int,java.lang.String[],int[])) and [this](https://developer.android.com/training/permissions/requesting#request-permission).

Also, it is better to check the permission in activity or fragment, but based on result of requesting permissions, call the appropriate function from viewmodel or do some another stuff.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+interface ContactActionListener {
+
+    fun onContactDelete(contact: Contact)
+}
```
Better extract to separate class.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+
+class ContactsAdapter(
+    private val userActionListener: ContactActionListener
+) : ListAdapter<Contact, ContactsAdapter.ContactsViewHolder>(ContactsDiffCallback()), View.OnClickListener {
```
You are not going to need to inherit from View.OnClickListener.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+        val binding = ItemContactBinding.inflate(inflater, parent, false)
+
+        binding.root.setOnClickListener(this)
+        binding.buttonTrash.setOnClickListener(this)
```
Get rid of these.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+            } else {
+                imageViewContactImage.setImageResource(R.drawable.default_contact_image)
+            }
+        }
```
In onBindViewHolder() function just call holder' bind() function.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+            }
+        }
+
+    }
```
Get rid of this.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+
+    class ContactsViewHolder(
+        val binding: ItemContactBinding
+    ) : RecyclerView.ViewHolder(binding.root)
```
Create bind() function and use it scope for configuring holder view. I.e. move old code from onBindViewHolder() to this function and also set on click listeners in this function. General rule for setting listeners in adapter: you should declare all instructions in activity or fragment class and pass them to adapter but use id of item from recycler adapter.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+    Glide.with(this).load(imageLink).apply(RequestOptions.circleCropTransform())
+        .placeholder(R.drawable.default_contact_image).error(R.drawable.default_contact_image)
+        .into(this)
+}
```
This is look like function for common util purposes. Extract her from adapter class to "util" class.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt:
```diff
@@ -0,0 +1,76 @@
+import com.alexP.assignment1.model.Contact
+import com.alexP.assignment1.viewModels.ContactsViewModel
+
+class ContactsLoader(private val context: Context, private val viewModel: ContactsViewModel) {
```
I think it is bad idea to pass viewmodel instance anywhere. Consult with mates in discord about using usecases here (because i do not know how to use it) or you can do all this stuff in viewmodel directly.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    private fun fetchContacts() {
+        val resolver: ContentResolver = context.contentResolver
```
It is better to pass instance of ContentResolver directly to fetchContracts() function parameters than pass it to class properties.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/model/ContactsService.kt:
```diff
@@ -0,0 +1,77 @@
+
+import com.github.javafaker.Faker
+
+typealias ContactsListener = (contacts: List<Contact>) -> Unit
```
Don't understand about the use of this typealias. Please explain at free time.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/AddContactFragment.kt:
```diff
@@ -0,0 +1,132 @@
+
+    interface OnContactSavedListener {
+        fun onContactSaved(contact: Contact)
+    }
```
Better extract to separate class.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/BaseActivity.kt:
It is better to extract this class to separate package

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt:
```diff
@@ -0,0 +1,170 @@
+import com.google.android.material.snackbar.Snackbar
+
+
+class ContactsActivity : AppCompatActivity(), AddContactFragment.OnContactSavedListener {
```
It's better to create a property with an interface type rather than inherit from it.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt:
```diff
@@ -0,0 +1,170 @@
+    private fun showAddContactDialog() {
+        val fragmentManager = supportFragmentManager
+        val fragment = AddContactFragment()
+        fragment.setOnContactSavedListener(this)
```
And then you can directly pass the implementation of the interface, and not the entire activity (or fragment).

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt:
```diff
@@ -0,0 +1,170 @@
+                .addToBackStack(null)
+                .commit()
+        }
+    }
```
Please explain me how it works (in discord voicechat at free time), i never seen this before.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/utils/ViewModelFactory.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+}
+
+fun Activity.factory() = ViewModelFactory(applicationContext as App)
```
Ideally, you should have a viewmodel for each activity or fragment to be ready execute business logic in them, but this ext function will be able create factory only for ContactsViewModel.
Also check out this [information](https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-cheatsheet).

Pl-Alex committed:
Fixed issues after review

CHANGED:
- code refactoring
FIXED:
- bug with fast deletion of contacts

Pl-Alex committed:
Architecture improvements and bug fixes
ADDED
- adjusted ContactsActivity for status and navigation bars and display cutouts
- AddContactViewModel with galleryUri StateFlow
- Improved architecture
FIXED
- crash during configuration changes in AddContactFragment
- validation refactored
REMOVED
- opening of the AddContactFragment dialog on a non-full screen on screens larger than or equal to xLarge

Pl-Alex committed:
Architecture improvements and bug fixes
ADDED
- adjusted ContactsActivity for status and navigation bars and display cutouts
- AddContactViewModel with galleryUri StateFlow
- Improved architecture
FIXED
- crash during configuration changes in AddContactFragment
- validation refactored
REMOVED
- opening of the AddContactFragment dialog on a non-full screen on screens larger than or equal to xLarge

1937Carrie commented on contactsprovider/data/src/main/java/com/alexp/contactsprovider/data/Contact.kt:
```diff
@@ -0,0 +1,12 @@
+    val email: String,
+    val phone: String,
+    val address: String,
+    val dateOfBirth: String
```
Check out [material](https://kotlinlang.org/docs/coding-conventions.html#trailing-commas) about trailing commas and reformat all your code.

1937Carrie commented on contactsprovider/data/src/main/AndroidManifest.xml:
Collapse empty tag. At the review, can tell your mentor that you optimized the file size by reducing it by 17,74% (62 B -> 51 B). Make it for all new manifests.

1937Carrie commented on datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt:
```diff
@@ -0,0 +1,74 @@
+        saveString(PASSWORD, password)
+    }
+
+    suspend fun saveBoolean(key: String, value: Boolean) {
```
Make all highlighted functions private.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:
```diff
@@ -0,0 +1,149 @@
+) : DialogFragment() {
+
+    private lateinit var binding: FragmentDialogAddContactBinding
+    private lateinit var viewModel: AddContactViewModel
```
Now that you've practiced creating a model instance manually, let's take a look at some other [ViewModel APIs](https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-cheatsheet) that make creating a model easier.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:
```diff
@@ -0,0 +1,149 @@
+        savedInstanceState: Bundle?
+    ): View {
+        binding = FragmentDialogAddContactBinding.inflate(inflater, container, false)
+        binding.root.applyWindowInsets()
```
It is good idea using extension function for displaying content edge-to-edge but a little bit wrong implementation. Remind me at you free time, I will show you what is wrong and how to handle it.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    private fun observeValues() {
+        lifecycleScope.launch {
```
Using this approach in fragments may cause to app crashing. Add following code, run program, come in and come out by using nav icon on toolbar from that fragment. Think about why the crash occurred.
```kotlin
override fun onDestroyView() {
        super.onDestroyView()
        i = null
    }

private var i: Int? = 0
private fun observeValues() {
    lifecycleScope.launch {
        while (true) {
            delay(250L)
            i = i!! + 1
            binding.topBar.title = i.toString()
        }
        viewModel.galleryUri.collect { uri ->
            uri?.let {
                binding.imageViewProfileImage.loadCircularImage(it.toString())
            }
        }
    }
}
```

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:
```diff
@@ -0,0 +1,149 @@
+
+        observeValues()
+        val toolbar = binding.topBar
+        toolbar.setNavigationOnClickListener {
```
Concatenate val toolbar declaration and setNavigationOnClickListener function call.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:
```diff
@@ -0,0 +1,149 @@
+        dismiss()
+    }
+
+    private fun enteredDataIsInvalid(): Boolean {
```
This function is performing asserting inputted string is valid. Things that are not related to UI layer have to be extracted to other layers and be calling from viewmodel.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt:
```diff
@@ -29,49 +29,34 @@ class MyProfileActivity : BaseActivity<ActivityMyProfileBinding>() {
 
-        binding?.let {
+        binding.let {
             Glide.with(this)
```
You already have function ImageView.loadCircularImage(). Replace Glide snippet with your extension function.

Pl-Alex committed:
Merge remote-tracking branch 'origin/task-2' into task-2

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt
#	app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt
#	app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactViewModel.kt
#	app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthActivity.kt
#	app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthViewModel.kt
#	app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt
#	app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsViewModel.kt
#	app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt
#	app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileViewModel.kt
#	contactsprovider/data/src/androidTest/java/com/alexp/contactsprovider/data/ExampleInstrumentedTest.kt
#	contactsprovider/data/src/main/AndroidManifest.xml
#	contactsprovider/data/src/test/java/com/alexp/contactsprovider/data/ExampleUnitTest.kt
#	datastore/data/src/androidTest/java/com/alexp/data/store/ExampleInstrumentedTest.kt
#	datastore/data/src/main/AndroidManifest.xml
#	datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt
#	datastore/data/src/test/java/com/alexp/data/store/ExampleUnitTest.kt
#	gradle/libs.versions.toml
#	textvalidation/data/src/androidTest/java/com/alexp/textvalidation/data/ExampleInstrumentedTest.kt
#	textvalidation/data/src/main/AndroidManifest.xml
#	textvalidation/data/src/main/java/com/alexp/textvalidation/data/validator/EmailValidator.kt
#	textvalidation/data/src/main/java/com/alexp/textvalidation/data/validator/EmptyValidator.kt
#	textvalidation/data/src/main/java/com/alexp/textvalidation/data/validator/PasswordValidator.kt
#	textvalidation/data/src/main/java/com/alexp/textvalidation/data/validator/base/IValidator.kt
#	textvalidation/data/src/test/java/com/alexp/textvalidation/data/ExampleUnitTest.kt

Pl-Alex committed:
Fixed AddContactFragment

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+        return oldItem == newItem
+    }
+
+}
```
I think it is better to move ContactsDiffCallback class implementation below ContactsAdapter class, because this Kotlin file is called ContactsAdapter, hense main object is ContactsAdapter. Related to entire project.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+    ) : RecyclerView.ViewHolder(binding.root) {
+        fun bind(contact: Contact) {
+            with(binding) {
+                itemView.tag = contact
```
For lines 48, 49, 68 and related to them, I think it is better to avoid using tags and implement clickListeners directly in setOnClickListener() function for each element respectively.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthActivity.kt:
Handling for password textInputLayout error description is missing.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthViewModel.kt:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    companion object {
+        fun createFactory(context: Context): ViewModelProvider.Factory = viewModelFactory {
```
I made a mistake using this approach. You should create a DataStoreProvider instance outside the factory and pass that instance to factory. So the createFactory() function should take a DataStoreProvider instance, not a Context. Related to entire project.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt:
```diff
@@ -0,0 +1,182 @@
+class ContactsActivity : BaseActivity<ActivityContactsBinding>() {
+    private lateinit var adapter: ContactsAdapter
+
+    private lateinit var viewModel: ContactsViewModel
```
Its time to get rid of lateinit var viewModel declaration and change it using viewModels() api, e.g. like you have already done in AddContactFragment. Related to entire project.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt:
```diff
@@ -0,0 +1,182 @@
+            }
+        }
+
+    private val onSaveAction: (Contact) -> Unit =
```
It is better change to style like
```kotlin
private val onSaveAction: (Contact) -> Unit = { contact: Contact ->
    viewModel.addContact(contact)
}
```

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt:
```diff
@@ -0,0 +1,61 @@
+
+    private lateinit var viewModel: MyProfileViewModel
+
+    companion object {
```
Move companion object to the bottom of class. [Source](https://kotlinlang.org/docs/coding-conventions.html#class-layout).

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt:
Handling for intent parameter(email) is missing.

1937Carrie reviewed (changes requested):
As a result of the architecture redesign, the display of errors in the password textInputLayout, and the receipt and processing of intent parameter have missed - need to bring back it. There are some style mistakes; mistake when working with recyclerview. Overall, it's fine.

Pl-Alex commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactViewModel.kt:
```diff
@@ -0,0 +1,66 @@
+import com.alexp.textvalidation.data.validator.base.ValidationResult
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+
```
q

Pl-Alex commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:
```diff
@@ -0,0 +1,153 @@
+        binding.inputLayoutUsername.error = validationResult?.let { getString(it) } ?: ""
+        return validationResult == null
+    }
+
```
q

Pl-Alex committed:
Autologin feature fixed
Validation on Auth and AddContact screens fixed
Added username in MyProfile

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -43,35 +30,40 @@ class ContactsAdapter(
         private val userActionListener: IContactActionListener,
     ) : RecyclerView.ViewHolder(binding.root) {
+
+        private lateinit var contact: Contact
```
Get rid of this.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
```diff
@@ -43,35 +30,40 @@ class ContactsAdapter(
 
+        fun getContact(): Contact {
+            return contact
         }
```
Get rid of this also.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:
Move whole content of package adapters to package contactsActivity, because this stuff is used only in the ContactsActivity.
<img width="248" alt="image" src="https://github.com/user-attachments/assets/af3b485e-29a3-4f9b-861c-e9c3f10c07ff">

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt:
```diff
@@ -113,7 +110,7 @@ class ContactsActivity : BaseActivity<ActivityContactsBinding>() {
 
             override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
-                deleteContact(viewHolder.itemView.tag as Contact)
+                deleteContact((viewHolder as ContactsAdapter.ContactsViewHolder).getContact())
```
Since I already said to remove the getContact() function from the ViewHolder, there is another way to accomplish deleting a contact from the list. In this particular case, the order of contacts in the viewmodel field is equal to the order in which contacts are displayed in the recyclerview.
Your code will be look like
```kotlin
val contact = /*retrieve contact using the default stack of fields or properties of passed viewholder*/
deleteContact(contact) // or little bit transformed form that respects contacts nullability
```

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactViewModel.kt:
It is bad you removed these functions from viewmodel. As main purpose of these functions is provide some business logic to UI, so these functions must be called from viewmodel since viewmodel is stateholder for UI. So yes, you have to called these function in viewmodel and then called the required viewmodel function in UI class. You're throwing data around between the UI, the viewmodel, and where that logic is defined, but this is the way.

Pl-Alex committed:
Corrections after review

Fіxed:
 - email text input in authActivity
 - returned validation functions to the viewmodels
 - password storage has been removed
 - Fixed an error in displaying the username
 - The approach to detecting a swiped contact has been changed

1937Carrie commented on datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt:
You misunderstood. Sign Up screen is imitation of authorization process before fifth level. But for authorization you need both an email and a password. So you either need store email and password together  if checkbox "Remember Me" was checked or don't store anything otherwise. A function saveRememberMeState() is redundant in this case. As a conclusion, when you start your app every next time, you check if datastore have credentials and direct the flow of code execution in the right direction, based on that result.
The behaviour have to be as in any of modern apps such telegram, instagram, gmail, etc. If datastore have credentials, you even do not see authorization screen, in ideally written project and you will immediately be taken to the main screen of the application.

1937Carrie replied:
And I think you should rework your DataStore read value functions so they will return Flow\<T\> but not T, because it is implicitly recommended in the [documentation](https://developer.android.com/topic/libraries/architecture/datastore#preferences-read). Now, if i understood correctly, you are writing value to datastore and waiting while value will be successfully written. But you need just write value to data store and wait only when you are reading the value.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt:
```diff
@@ -110,7 +108,9 @@ class ContactsActivity : BaseActivity<ActivityContactsBinding>() {
 
             override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
-                deleteContact((viewHolder as ContactsAdapter.ContactsViewHolder).getContact())
+                val contact = vm.contacts.value
+                    ?.get((viewHolder as ContactsAdapter.ContactsViewHolder).adapterPosition)
+                contact?.let{ deleteContact(contact)}
```
The following code will be enough:
```kotlin
val contact = vm.contacts.value?.get(viewHolder.adapterPosition)
contact?.let{ deleteContact(contact)}
```

Pl-Alex committed:
Corrections after review

Fіxed:
 - Authorization logic fixed
 - Read value functions from a DataStore now return flows
 - Refactored some code for better readability

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthViewModel.kt:
```diff
@@ -23,25 +24,30 @@ class AuthViewModel(
+    fun saveUsername(username: String) {
         viewModelScope.launch {
-            dataStore.saveCredentials(email)
-            dataStore.saveRememberMeState(isRememberMeChecked)
-            updateAuthState()
+            dataStore.saveUsername(username)
         }
     }
```
Redundant code

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthActivity.kt:
```diff
@@ -64,9 +65,15 @@ class AuthActivity : BaseActivity<ActivityAuthBinding>() {
+        val passwordText = binding.inputEditTextPassword?.text.toString()
+        val username = parseEmail(emailText)
+
+        vm.saveUsername(username)
```
Redundant code.

1937Carrie commented on app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileViewModel.kt:
```diff
@@ -15,13 +15,15 @@ class MyProfileViewModel(
         viewModelScope.launch {
-            _emailState.update {
-                dataStore.readEmail()
+            dataStore.readUsername().collect { username ->
+                _myProfileState.update {
+                    MyProfileState(username = username)
+                }
```
Replace readUsername() function with readEmail() and format that string like
```kotlin
val formattedName = email.substringBefore("@").split(".").joinToString(" ")
```

1937Carrie commented on datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt:
```diff
@@ -3,69 +3,65 @@ package com.alexp.datastore.data
-private const val EMAIL = "email"
-private const val REMEMBER_STATE = "remember_state"
+private val emailKey = stringPreferencesKey("email")
+private val usernameKey = stringPreferencesKey("username")
```
Get rid of this.
